api/python/provisioner/lock.py

```diff
@@ -0,0 +1,229 @@
+    _lock_filename_tmpl = Template("provisioner.$TARGETS.lock")
+
+    @staticmethod
+    def _check_pid(pid: str, target: str) -> bool:
+        """
+        Check if process is running by given pid and on given salt target.
+
+        Parameters
+        ----------
+        pid: str
+            process pid for check
+        target: str
+            target where need to verify the pid of a process
+
+        Returns
+        -------
+        bool:
+            True if process by given `pid` is alive on `target`.
+            False otherwise
+
+        """
+        # run the command: `salt $target ps.kill_pid $pid signal=0`
+        # NOTE: we send 0 signal to check for the existence of a process ID
+        res = function_run('ps.kill_pid', fun_args=[pid],
+                           fun_kwargs=dict(signal=0),
+                           targets=target)
+        return bool(res[target])
+
+    def _get_lock_files(self) -> Iterable[Path]:
+        """
+        Get the provisioner lock files
+
+        Returns
+        -------
+        iter:
+            Iterator object over all provisioner lock files in the lock dir
+        """
+        # NOTE: search for files by pattern 'provisioner.*.lock'
+        return Path(PRVSNR_LOCKS_FILES_DIR).glob(
+            self._lock_filename_tmpl.substitute(TARGETS="*"))
+
+    def _handle_old_lock_files(self):
+        """
+        Method handles all existing provisioner lock files:
+
+        1. Checks if they have the correct metadata and delete if don't
+        2. If lock file metadata is not actual, delete it
+
+        Returns
+        -------
+        None
+
+        """
+        exc_ctx = None
+        for lock_file in self._get_lock_files():
+            try:
+                lock_metadata = utils.load_json(lock_file)
+            except json.decoder.JSONDecodeError:
+                # Metadata is not well-formatted as proper json
+                pass  # delete this lock file in the end of the loop
+            else:
+                if lock_metadata:
+                    target = lock_metadata.get(
+                        LockMetaDataFields.SOURCE_TARGET.value, None)
+                    pid = lock_metadata.get(
+                        LockMetaDataFields.PID.value, None)
+                    if (target and pid and
+                            check_salt_minions_are_ready(targets=[target]) and
+                            self._check_pid(pid, target)):
+
+                        exc_ctx = LockContext(lock_file=lock_file, pid=pid,
+                                              targets=target)
+                        # NOTE: to clean all orphan lock files
+                        continue  # raise an exception later
+
+            # NOTE: Delete file in the following cases:
+            #  1. Necessary metadata fields are missed
+            #  2. The lock file initiator target is not alive
+            #  3. The process by given pid is not running
+            lock_file.unlink()
```

Comment:
please log warning here since we are removing some lock file, it would be greate to have a reason as well (malformed, empty, orphan ...)

Reply:
I agree, it makes a sense. I added logger warning message here with appropriate reason message